python/argo/interaction.py

```diff
@@ -6,21 +6,33 @@
 from typing import Any, Dict, Tuple
 from typing_extensions import Protocol
 
+
 class HasProtocolState(Protocol):
     def protocol_state(self) -> Any: ...
-    server_connection : ServerConnection
+    server_connection: ServerConnection
 
-class ArgoException(Exception):
-    message : str
-    data : Dict[str, Any]
-    code : int
 
+class ArgoException(Exception):
+    message: str
+    data: Dict[str, Any]
+    code: int
+    stdout: str
+    stderr: str
```

Comment:
I think str is the right thing here, but I want to be sure. What do we do if something dumps invalid UTF-8 (e.g. some literal bytes from some LLVM files)? Will the Haskell library behind the scenes catch it?

Reply:
Haskell `String`s are guaranteed to be valid UTF-8, aren't they? I would have to test this to be sure, but *something* ought to break and throw an exception—which would then be caught, wrapped, and returned to the client.

Reply:
Haskell Strings are not UTF-8, because they represent code points rather than bytes. But I don't think we can have a list of Haskell Chars that spits out invalid UTF-8 for Python to slurp in.

Reply:
Nothing says that stdio has characters in it on Unix, though - it's just bags of bytes, and UTF-8 is a convention. But the Haskell library that you're using to capture stdio takes care of this, right?